### DISCUSSION TITLE
Security concerns with jwt.InferAlgorithmFromKey(true)

### DISCUSSION THREAD

**Discussion by jordaniversen** [lestrrat-go/jwx#643] [General]
Hello,

I'm just curious as to whether the warning at https://github.com/lestrrat-go/jwx/blob/main/docs/01-jwt.md?plain=1#L184 is still valid.  By looking at the code at https://github.com/lestrrat-go/jwx/blob/main/jwt/jwt.go#L282 it seems like it is simply checking for an exact match of the algorithm, and not attempting to try multiple algorithms in a brute force way.

Please forgive any ignorance, thank you!

**Comment by lestrrat** [comment]
Not sure where you are getting this from

>  it seems like it is simply checking for an exact match of the algorithm, and not attempting to try multiple algorithms in a brute force way.

There is a for loop there.

https://github.com/lestrrat-go/jwx/blob/b88feb74d96610906155732f429b5fd3b7ef2ae6/jwt/jwt.go#L274-L288

**Comment by jordaniversen** [comment]
Hi

Thanks for the quick response.  Ya I saw the for loop but still isn't it only going to do a single action against a single algorithm that exactly matches?  I interpreted that documentation to mean that it will try to verify the signature in multiple ways or with multiple algorithms and "hope" that one succeeds or something, so perhaps I just misinterpreted it?  Sorry if this is such a trivial question, I just wanted to confirm this case as if it is unsafe to rely on this then I unfortunately need to try combining multiple other libraries, which I dont want to do because this one seems much more flexible and extensible.

**Comment by lestrrat** [comment]
I think I see where you're coming from. to give you some history, the AlgorithmsFor() bit was slapped on top of what it used to do to make _keys_ that don't have `alg` to work... so I think that's why it's the way it is. I think there _is_ discrepancy between what was intended and what the implementation is:

I think what should have been done is (pseudocode)

````
for alg in algs {
   if token.alg == "" || token.alg == alg {
      tryVerify
    }
}
```

This way it works for cases where keys without alg are used but the token itself has one, or the case when the token itself doesn't have one. Does this seem like a better approach?

**Comment by jordaniversen** [comment]
Ah if it is simply a slight inefficiency then I'm ok with that, but if it
was a security issue that I wasn't seeing then that's a different story. :)

Thanks

**Comment by lestrrat** [comment]
Oh wait, it _is_ written that way (spoke from memory)

```
  for _, alg := range algs {
    // if we have a "alg" field in the JWS, we can only proceed if
    // the inferred algorithm matches
    if hdrAlg != "" && hdrAlg != alg {
      continue
    }
 ```

**Comment by lestrrat** [comment]
Well, there _is_ a security issue, no? I mean, if the alg field in the token or key was empty, then how do you know if the verification that succeeded was a result of an unintended bug or not (maybe they meant to rotate RS256+an old key to  RS512+new key): You can't tell if you're doing the right thing, or they are still using the old key by mistake. Matching the algorithm is a security concern. I personally sympathize with the programmers that need to deal with such a situation, but I don't with those who made the decision to omit the algorithm

**Reply by jordaniversen** [reply]
I'm not sure if I'd use the argument about there being an unintended bug--that can be applied to any part of any software, not particular to this discussion IMHO.

Is there a way (or can it be added) to register something like "allowed algorithms"?  That way we can make sure we don't register something like the "none" and "HMAC" algorithms (obviously, for the use cases where those are actually not expected), to prevent a hacker tampering with the "alg" in the header:
https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/#Recommendations-for-Library-Developers

In my case, I cannot control the JWKS and they do not have "alg" in the JWK, and they only use ECDSA.

Thanks

**Reply by lestrrat** [reply]
> I'm not sure if I'd use the argument about there being an unintended bug--that can be applied to any part of any software, not particular to this discussion IMHO.

I firmly stand in the stance that the JWS/JWK RFC is flawed for including and/or not making `alg`'s use more well-defined...  It's a security issue in the protocol, which the users need to be aware of in order to not be bitten themselves. And yes, the auth0 article has been around for at almost 7 years, trust me, I know.

> Is there a way (or can it be added) to register something like "allowed algorithms"?

My current answer is no. As a concept it's possible, but I've never come up with a good user API for this. How are you going to specify algorithms for JWKS (i.e. multiple keys) in the API call?

Pass a map of `kid` to list of algorithms that a particular key can use?

```go
// v1
jws.Verify(signed, jws.WithKeySet(set), <a map of ... kid to algorithms?>)

// v2
jws.Verify(signed, jws.WithKeySet(set, <again... a map??>))
```

Or, just allow a single set of algorithms for the entirety of the JWKS?

```go
// v1
jws.Verify(signed, jws.WithKeySet(set), jws.WithAlgorithms(...))

// v2
jws.Verify(signed, jws.WithKeySet(set, jws.WithAlgorithms(...)))
```

Everything I've come up so far over the years looked clunky. Ideas are welcome.

> In my case, I cannot control the JWKS and they do not have "alg" in the JWK, and they only use ECDSA.

You can do whatever you want to JWKS in your code, no? (which also happens to be one of the reasons this library does the entirety of JOSE, not just JWTs)

```go
set, _ := jwk.Fetch(...)
// You already know you can only use ECDSA, right? Just set the algorithm
// that you wish to use.
for i := 0; i < set.Len(); i++ {
  key, _ := set.Get(i)
  key.Set(jwk.AlgorithmKey, jwa.ES256) // or whatever
}
```

in `v2`, you can do this in the `jws.KeyProvider`:

```go
jws.Verify(signed, jws,WithKeyProvider(jws.KeyProviderFunc(func(ctx context.Context, sink jws.KeySink, sig *jws.Signature, msg *jws.Message ) error {
   // load your set
   set, _ := jwk.Fetch(...)
   
   for i := 0; i < set.Len(); i++ {
     key, _ := set.Get(i)
     sink.Key(jwa.ES256, key)
  }
  return nil
}))
```

**Reply by jordaniversen** [reply]
`  key.Set(jwk.AlgorithmKey, jwa.ES256) // or whatever`

Thanks.  In my case, i'm using `jwt.WithKeySetProvider` and autorefresh.  So I guess with this approach, I would need to put that for loop and hard coding of the algorithm in my `jwt.KeySetProviderFunc` function meaning it would need happen each time I need to parse a JWT, correct?  I.e. I couldn't do it once, outside of `jwt.KeySetProviderFunc` due to the auto refresh, I assume.  If this is correct, what would be the best way around this (or would it be a future improvement)?  I assume some callback to define once auto refresh has completed.  And yes I agree the repeating "for" loop calls really isn't the end of the world in terms of performance.

**Reply by lestrrat** [reply]
> I assume some callback to define once auto refresh has completed. 

Ah ok. That's something that doesn't exist as of today, and that's definitely something we could do

**Reply by lestrrat** [reply]
done for v2 #650

**Reply by jordaniversen** [reply]
Awesome thanks will give it a try!

**Reply by jordaniversen** [reply]
Hi again @lestrrat,

I realize this isnt an OIDC library, but as per https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys is there a way to auto refresh the keys in the JWKS if a "kid" isn't found?  If not, I guess that makes sense because again this isnt OIDC, and I'll use the auto refresh internal.
